src/fluree/db/ledger/transact/validation.clj

```diff
@@ -0,0 +1,599 @@
+                                               f (update :queue conj f)
+                                               true (update-in [:tx-spec pid-or-sid] into flakes))))
+    :collection (swap! validate-fn (fn [validate-data]
+                                     (let [existing-flakes (get-in validate-data [:c-spec pid-or-sid])]
+                                       (cond-> validate-data
+                                               true (assoc-in [:c-spec pid-or-sid] (into existing-flakes flakes))
+                                               ;; if flakes exist, we already queued a fn for this subject, don't want multiple
+                                               (empty? existing-flakes) (update :queue conj f)))))))
+
+
+;; TODO - this can be done per schema change, not per transaction, to make even more efficient.
+(defn build-function
+  "Builds a function based on a function subject-id (or a list of function subject ids) and
+  delivers the executable function to the provided promise.
+
+  Promise is used here to cache this work so it is only done once per transaction.
+
+  fn-type should be:
+  - predSpec - for predicate Spec
+  - "
+  [fn-subjects db promise fn-type]
+  (async/go
+    (try
+      (let [fn-str    (->> fn-subjects                      ;; combine multiple functions with an (and ...) wrapper
+                           (map #(query-range/index-range db :spot = [% const/$_fn:code]))
+                           async/merge
+                           (async/into [])
+                           (<?)
+                           (map #(if (util/exception? %) (throw %) (.-o ^Flake (first %))))
+                           dbfunctions/combine-fns)
+            fn-parsed (<? (dbfunctions/parse-fn db fn-str fn-type nil))]
+        (deliver promise fn-parsed))
+      (catch Exception e (deliver promise e)))))
+
+
+(defn resolve-function
+  "Returns promise for function that will get used. Tries to get one from cache
+  and if not already processed then will create a new promise channel.
+  Resistant to race conditions which are unlikely, but possible."
+  [fn-subjects db validate-fn-atom fn-type]
+  (or (get-in @validate-fn-atom [:cache fn-subjects])
+      (let [p (promise)]
+        (swap! validate-fn-atom update :cache
+               (fn [fn-cache]
+                 (if (get fn-cache fn-subjects)
+                   fn-cache                                 ;; something put a promise in here while we were checking, just return
+                   (do (build-function fn-subjects db p fn-type)
+                       (assoc fn-cache fn-subjects p)))))
+        ;; return whatever promise was in the cache - either one we just created or existing one if a race condition existed
+        (get-in @validate-fn-atom [:cache fn-subjects]))))
+
+
+(defn- async-response-wrapper
+  "Wraps an async response with a formatted response."
+  [async-response function]
+  (async/go
```

Comment:
i think this can be:

```clojure
(defn- async-response-wrapper
  [async-response function]
  (async/pipe async-response
              (chan 1 (map function)))
```

Reply:
Cool alternative, putting in place.